packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt

```diff
@@ -23,17 +24,48 @@ internal class ImageLibraryContract(
     Intent(Intent.ACTION_GET_CONTENT)
       .addCategory(Intent.CATEGORY_OPENABLE)
       .setType(input.options.mediaTypes.toMimeType())
+      .apply {
+        if (input.options.allowsMultipleSelection) {
+          this.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
+        }
+      }
 
   override fun parseResult(input: ImageLibraryContractOptions, resultCode: Int, intent: Intent?) =
     if (resultCode == Activity.RESULT_CANCELED) {
       ImagePickerContractResult.Cancelled()
+    } else if (input.options.allowsMultipleSelection) {
+      val uris = requireNotNull(intent).getClipDataUris()
+      val contentResolver = requireNotNull(appContextProvider.appContext.reactContext) { "React Application Context is null. "}.contentResolver
+      ImagePickerContractResult.Success(uris.map { uri -> uri.toMediaType(contentResolver) to uri })
     } else {
       val uri = requireNotNull(requireNotNull(intent).data)
       val contentResolver = requireNotNull(appContextProvider.appContext.reactContext) { "React Application Context is null. " }.contentResolver
       val type = uri.toMediaType(contentResolver)
+      ImagePickerContractResult.Success(listOf(type to uri))
+    }
+}
 
-      ImagePickerContractResult.Success(type to uri)
+/**
+ * Copied from [androidx.activity.result.contract.ActivityResultContracts.GetMultipleContents.getClipDataUris]
+ */
+internal fun Intent.getClipDataUris(): List<Uri> {
+  // Use a LinkedHashSet to maintain any ordering that may be present in the ClipData
+  val resultSet = LinkedHashSet<Uri>()
+  data?.let { data ->
+    resultSet.add(data)
+  }
+  val clipData = clipData
+  if (clipData == null && resultSet.isEmpty()) {
+    return emptyList()
+  } else if (clipData != null) {
+    for (i in 0 until clipData.itemCount) {
+      val uri = clipData.getItemAt(i).uri
+      if (uri != null) {
+        resultSet.add(uri)
+      }
```

Comment:
Maybe you'll take some inspiration from:
https://github.com/barthap/expo-music-picker/blob/536168cc1e8ec88371d0dad3aad2833773b82df8/android/src/main/java/expo/community.modules.musicpicker/Utilities.kt#L64-L75
And then you can
```kotlin
for (item in clipData.items) {
 resultSet.add(item.uri)
}